Document how clients not on K8s can join a DC in K8s (#9438)
hashicorp/consul: website/content/docs/k8s/installation/deployment-configurations/clients-outside-kubernetes.mdx

Consul clients running on non-Kubernetes nodes can join a Consul cluster running within Kubernetes.

## Networking

Within one datacenter, Consul typically requires a fully connected
[network](/docs/architecture). This means the IPs of every client and server
agent should be routeable by every other client and server agent in the
datacenter. Clients need to be able to gossip with every other agent and make
RPC calls to servers. Servers need to be able to gossip with every other

> agent.

Reviewer: ```suggestion
agent. See [Architecture](/docs/architecture) for more details.
```

  2.  On the external VM, run:
  ```bash
  consul agent \
    -advertise="$ADVERTISE_IP" \
    -retry-join='provider=k8s label_selector="app=consul,component=server"'
    -bind=0.0.0.0 \
    -client=0.0.0.0 \
    -hcl='leave_on_terminate = true' \
    -hcl='ports { grpc = 8502 }' \
    -config-dir=$CONFIG_DIR \
    -datacenter=$DATACENTER \
    -data-dir=$DATA_DIR \
  ```

  3. Check if the join was successful by running `consul members`. Sample output:
  ```shell-session
  / $ consul members
  Node                                           Address           Status  Type    Build  Protocol  DC   Segment
  consul-consul-server-0                         10.138.0.43:9301  alive    server  1.9.1  2         dc1  <all>
  external-agent                                 10.138.0.38:8301  alive    client  1.9.0  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-grs4  10.138.0.43:8301  alive    client  1.9.1  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-otge  10.138.0.44:8301  alive    client  1.9.1  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-vo7k  10.138.0.42:8301  alive    client  1.9.1  2         dc1  <default>
  ```

> ### Auto-join on via host ports

Reviewer: ```suggestion
### Auto-join via host ports
```

a named port to specify the port for the auto-join to return. This enables

> different pods to have different exposed ports.

Author: I did not need to add this annotation, even though I configured my server gossip port to be 9301. `consul members` still successfully showed all members of the cluster.

Author: I think it's still useful documentation to have, but that's why I didn't include this annotation in the instructions below.

Reviewer: weird, I don't know how the join could have worked without this? Wouldn't it try 8301 unless explicitly configured?

Author: @ishustava I did change the gossip port on the server to 9301. I know its super weird I would expect it to use that code as well!

I would have expected the external client agent fail to join because it would try 8301. However, when trying the exactly the steps [here](https://github.com/hashicorp/consul-helm/pull/740) while replacing the retry-join with `-retry-join 'provider=k8s host_network=true label_selector="app=consul,component=server"'` I observed it joining just fine and on port 9301, which is definitely mind boggling. I will retry those steps one more time tomorrow morning just to make sure.

Author: Update: I ran through the steps with `-retry-join 'provider=k8s host_network=true label_selector="app=consul,component=server"'` on the external agent without the annotation and it works. In fact, when I do add the annotation, it doesn't accept an integer value, and if I add the port "9301" as a string, the consul clients on k8s itself fail to join saying no consul servers found. 

I think for these docs I should remove this annotation, since it doesn't seem to work. wdyt @ishustava @lkysow?

Reviewer: As discussed lets:
1. Add a warning that if you're using 9301 and your consul servers are on different nodes (which is unlikely because in that case you wouldn't need to use 9301) then you can't use autojoin
1. Create a bug in go-discover about the annotation issue

Author: I removed the note about the annotation since its currently broken (we can add it back once the bug is fixed in go-discover), and added the note/warning about default port. I phrased it a bit differently, let me know if you think its clear or not

Author: Writing the bug right now

a named port to specify the port for the auto-join to return. This enables
different pods to have different exposed ports.

### Auto-join on the Pod network

> In the default Consul Helm chart installation, the Consul clients and servers
> are deployed as pods and routeable only via their `podIP:containerPort` for
> gossip/RPC. This would mean any external client agents joining would need to
> be able to have connectivity to those pod IPs.

Reviewer: ```suggestion
In the default Consul Helm chart installation, Consul clients and servers
are routable only via their pod IPs for
gossip and server RPC (technically, HTTP API calls to Consul clients can also be made through host IPs). This means any external client agents joining the Consul cluster running on Kubernetes would need to
be able to have connectivity to those pod IPs.

In many hosted Kubernetes environments, you will need to explicitly configure your hosting provider to ensure that pod IPs are routable from external VMs. See [Azure AKS CNI](https://docs.microsoft.com/en-us/azure/aks/concepts-network#azure-cni-advanced-networking), [AWS EKS CNI](https://docs.aws.amazon.com/eks/latest/userguide/pod-networking.html) and [GKE VPC-native clusters](https://cloud.google.com/kubernetes-engine/docs/concepts/alias-ips).
```

* I think saying "in the default install consul clients and servers are deployed as pods..." might be a bit confusing because it sounds like there's another install where they're not deployed as pods.
* added a bit more colour but feel free to edit

Author: Love it!

Given you have the [official Helm chart](/docs/k8s/helm) installed with the default values, do the following to join an external client agent. 

1. Make sure the pod IPs of the clients and servers in Kubernetes are routeable from the VM and that the VM IP (`$ADVERTISE_IP`) is routeable from the pods, adding any firewall rules necessary to allow traffic on the gossip and RPC ports.

2. Make sure you have the `kubeconfig` file for the Kubernetes cluster in `$HOME/.kube/config`.

2.  On the external VM, run: 

> ```shell-session

Reviewer: This needs to be indented so it's part of the bullet above. Sometimes the docs formatter will unindent on commit unless it's indented 3 spaces.

consul agent \
  -advertise="$ADVERTISE_IP" \
  -retry-join='provider=k8s label_selector="app=consul,component=server"'
  -bind=0.0.0.0 \

>   -client=0.0.0.0 \

Reviewer: I don't think we want to recommend this because it opens up the HTTP api and because it's not required for the client to join the cluster.

Reviewer: bump on this

>   -hcl='leave_on_terminate = true' \

Reviewer 2: I think we can omit it here because it's true by default for clients. We are setting it explicitly in the helm chart to support earlier versions of consul that defaulted to false, but I don't think we need it for these docs.

>   -hcl='ports { grpc = 8502 }' \

Reviewer: This defaults to 8502 anyway so I don't think we need this

Reviewer: this too

Reviewer: ahh I'm incorrect! Ignore
> Default -1 (disabled).

  -config-dir=$CONFIG_DIR \

>   -datacenter=$DATACENTER \

Reviewer: ```suggestion
  -datacenter=$DATACENTER \ # Use the same datacenter name as your Kubernetes cluster
```

  -data-dir=$DATA_DIR \
```

> 3. Check if the join was successful by running `consul members`

Reviewer 2: I wonder if it makes sense to include a sample output of `consul members` if you have it somewhere still. If not, then don't worry about it!

Author: Yup I've got some! I like it!

> ### Auto-join on the Host network

Reviewer 2: I think since host network and host ports are two separate concepts in Kubernetes, it's worth being precise here since using host network has security implications.

```suggestion
### Auto-join via host ports
```

> If you don't have connectivity to the `podIP:containerPort` for clients and servers deployed in Kubernetes, you have the option to expose the clients and servers on the host network. 

Reviewer 2: ```suggestion
If you don't have connectivity to the `podIP:containerPort` for clients and servers deployed in Kubernetes, you have the option to expose the clients and servers ports on the host IP instead. 
```

Reviewer: Maybe say more explicitly "If your external VMs can't connect to Kubernetes pod IPs... but they _can_ connect to the host IPs of the nodes in the Kubernetes cluster then you have the option..."

Author: Love both the suggestions!

1. Install the [official Helm chart](/docs/k8s/helm) with the following values:

> ```yaml

Reviewer 2: I think we need to add indentation here so that the automatic numbering will work (and above too).

Author: Yup, thank you!!

client:
  exposeGossipPorts: true # exposes client gossip ports as hostPorts
server:
  exposeGossipAndRPCPorts: true # exposes the server gossip and RPC ports as hostPorts
  ports:
    # Configures the server gossip port
    serflan:
      # Note that this needs to be different than 8301, to avoid conflicting with the client gossip hostPort
      port: 9301 
```
 This will expose the client gossip ports, the server gossip ports and the server RPC port at `hostIP:hostPort`. Note that `hostIP` is the **internal** IP of the VM that the client/server pods are deployed on.

> 2. Make sure the pod IPs of the clients and servers in Kubernetes are routeable from the VM and that the VM IP (`$ADVERTISE_IP`) is routeable from the pods, adding any firewall rules necessary to allow traffic on the gossip and RPC ports for the K8s Nodes and the external VM.

Reviewer: ```suggestion
1. Make sure the IPs of the Kubernetes nodes are routable from the VM and that the VM can access ports 8301 and 9301 (for gossip) and port 8300 (for server RPC) on those node IPs.
1. Make sure the the client and server pods running in Kubernetes can route to the VM's advertise IP on its gossip port (default 8301).
```
Let's split the requirements out into two bullets, the first from VM -> Kube and the second from Kube -> VM

Author: Yup that sounds good. In this specific edit I want to say `Make sure the IPs of the Kubernetes **pods** are routable` rather than nodes (since this section is pod network).

Author: Sounds great, and thanks for catching that this needed to be edited for host network!

3. Make sure you have the `kubeconfig` file for the Kubernetes cluster in `$HOME/.kube/config`.

4. On the external VM, run (note the addition of `host_network=true` in the retry-join argument): 
```shell-session
consul agent \
  -advertise="$ADVERTISE_IP" \
  -retry-join='provider=k8s host_network=true label_selector="app=consul,component=server"'
  -bind=0.0.0.0 \
  -client=0.0.0.0 \
  -hcl='leave_on_terminate = true' \
  -hcl='ports { grpc = 8502 }' \
  -config-dir=$CONFIG_DIR \
  -datacenter=$DATACENTER \
  -data-dir=$DATA_DIR \
``` 
3. Check if the join was successful by running `consul members`

> ## Manual join

Reviewer 2: should we recommend to use the host port option for the manual join since the pod IPs could change during upgrades?

Reviewer: Yeah maybe instead let's recommend setting up a DNS entry that would resolve to the server's pod IPs (if using pod network) or node ips that the server pods are running on (if using host ports)?

Author: Yes that makes sense, thank you both. I've included that. Also, it doesn't need every consul server, right? I'm going to take out that phrase and say "pass the address of at least one consul server".
